### PR TITLE
KAFKA-5818: KafkaStreams state transitions not correct

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
PR for 0.11.0 \cc @dguy @guozhangwang @bbejeck

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 1]
```diff
@@ -277,6 +276,24 @@ private boolean setState(final State newState) {
...
+            state = State.RUNNING;
+            stateLock.notifyAll();
+        }
+
+        // we need to call the user customized state listener outside the state lock to avoid potential deadlocks
+        if (stateListener != null) {
+            stateListener.onChange(State.RUNNING, State.CREATED);
+        }
+
+        return true;
```
Comment (tedyu) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: There is only one value returned.
This method can be declared as void.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java]: Ack.

**Comment by mjsax** [comment]
JKD9 does only work on `trunk` -- so failing test is unrelated. Updated this.

**Comment by guozhangwang** [comment]
LGTM. Merged to 0.11.0.

**Comment by dguy** [comment]
Thanks @mjsax & @guozhangwang

File: streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java [hunk 2]
```diff
@@ -229,29 +230,29 @@ public void testCloseIsIdempotent() throws Exception {
...
+    @Test
     public void testCannotStartOnceClosed() throws Exception {
         streams.start();
         streams.close();
         try {
             streams.start();
-        } catch (final IllegalStateException e) {
-            Assert.assertEquals("Cannot start again.", e.getMessage());
-            throw e;
+            fail("Should have thrown IllegalStateException.");
```
Comment (viktorsomogyi) [inline streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:239]: Hi Matthias,
I think overall using `@Rule` annotations is a cleaner approach compared to the proposed solution - unless it's not applicable here for some reason I don't see.
More precisely I'm thinking about using ExpectedException like this:
```
 public class SimpleExpectedExceptionTest {
     @Rule
     public ExpectedException thrown = ExpectedException.none();

     @Test
     public void throwsNothing() {
         // no exception expected, none thrown: passes.
     }

     @Test
     public void throwsExceptionWithSpecificType() {
         thrown.expect(NullPointerException.class);
         throw new NullPointerException();
     }
 }
```
What do you think?

Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:239]: I guess that would work too. ATM, we use the "fail-pattern" throughout the whole code base. \cc @dguy WDYT? Should we start to transit to a "rule-pattern"?

**Pull request closed without merging**